src/libexpr/include/nix/expr/parser-state.hh

```diff
@@ -269,10 +276,35 @@ ParserState::stripIndentation(const PosIdx pos, std::vector<std::pair<PosIdx, st
                 if (str->p[j] == ' ')
                     curIndent++;
                 else if (str->p[j] == '\n') {
+                    /* if curIndent is less than the current value of minIndent,
+                     * we can't calculate at this point how much indention we
+                     * will remove; we will just have to over-allocate later.
+                     * Fortunately in practice this shouldn't come up very much.
+                     * It would have to look like (with spaces replaced with
+                     * underscore for clarity)
+                     *
+                     * ''
+                     * ________only long indentations initially
+                     * ________more long indentations
+                     * ____
+                     * ^^ empty line with a shorter indentation
+                     * ____after that there can be shorter indentations
+                     * ''
+                     *
+                     * most empty lines will have no indentation, and those that
+                     * do will usually be of at least the minium indentation,
+                     * and come after a line with said minimum indentation, in
+                     * which case we can perfectly pre-allocate the string.
+                     */
```

Comment:
I am not opposed to any of this stuff, per se, but I am a bit confused.

Would it work to constraint stuff on the regular heap while we are figuring it out, and then perfectly allocate the resulting final string in the arena? Or is this all more complicated than I think, and there is still some "unknowable" stuff after parsing about this indentation business?

(We could also imagine a second arena just for the lifetime of the `stripIndentation` call, instead of the regular heap, for the temporary stuff.)

Reply:
I would say since you have the temporary `std::vector` below, you are already making temporaries, so I would maybe just do that for now, if it means simpler code / smaller diff / etc.

Still worth looking into avoiding the heap temporaries eventually, but I would say that is lower priority than the rest of your Grand Plan. :)

Reply:
Fair enough. I double checked to be sure, and it doesn't affect performance noticeably, if at all. I appreciate the reality check :laughing: